[ Enhancement ] Standard API response in document module

jerome-white opened the pull request:
## Summary

Target issue is #120 

Updates document routes with standard API response. In doing so it also:

* Updates route test cases to be response compliant
* Adds the document migration file so that the routes actually work (#122)
* Corrects the test case naming collision that prevents tests from running (#123)

## Checklist

Before submitting a pull request, please ensure that you mark these task.

- [x] Ran `fastapi run --reload app/main.py` or `docker compose up` in the repository root and test.
- [x] If you've fixed a bug or added code that is tested and has test cases.

## Notes

The sub-issues assigned to #120 came about as development on this PR began: having to fix those was not known beforehand. As a result, they are baked in to this PR instead of being their own standalone PRs, which would have been the ideal approach.


---

This pull request introduces several enhancements to the document module in the `ai-platform` repository. Key changes include:

1. **Database Migration**: A new migration script is added to create a 'document' table. This table includes fields for id, owner_id, filename, object store URL, creation timestamp, and supports soft deletion.

2. **API Route Refactoring**: Document handling routes are refactored to utilize the `APIResponse` wrapper, which standardizes HTTP status codes and improves error message formatting. This change enhances error handling and updates response models for consistency.

3. **CRUD Operations Simplification**: The document CRUD operations are simplified by removing the `DocumentList` model wrapper. Additionally, error message formatting for ownership validation is improved.

4. **Model Import Simplification**: Document model imports are streamlined by removing `DocumentList` and consolidating to a single `Document` import.

5. **Document Model Update**: The `Document` model is updated with revised foreign key constraints and the removal of the `DocumentList` class. String quote usage is standardized, and a back-populated relationship with the `User` model is added.

6. **Utility Function Enhancement**: The `failure_response` method in the `APIResponse` class is made more flexible by allowing the error parameter to accept any type, which is then converted to a string.

These changes collectively aim to enhance the standardization and flexibility of the document module's API responses.

jerome-white committed:
Remove DocumentList

DocumentList was used as the return type of multiple Documents because
of the precedent set in the original platform code. The importance of
having this type, and not just using a native list of Document's, is
not clear.

jerome-white committed:
Allowing the method to be type agnostic

jerome-white committed:
Update document routes to use standard API response

jerome-white committed:
Rename document test cases to support automatic testing

jerome-white committed:
Use non-standard response for erroneous cases

The standard API response does not raise an exception on failures, so
all responses are reflected to be "success".

jerome-white committed:
Convenience method for converting HTTP response to API-standard

jerome-white committed:
Test cases are APIResponse aware

jerome-white committed:
Document database tables

jerome-white committed:
Revert failure_response definition

jerome-white committed:
Dead return statement

jerome-white committed:
Invalid variable reference

jerome-white commented on backend/app/alembic/versions/c43313eca57d_add_document_tables.py:
```diff
@@ -0,0 +1,41 @@
+        ),
+        sa.Column("created_at", sa.DateTime(), nullable=False),
+        sa.Column("deleted_at", sa.DateTime(), nullable=True),
+        sa.ForeignKeyConstraint(["owner_id"], ["user.id"], ondelete="CASCADE"),
```
This is not appropriate to fix in this PR

jerome-white commented on backend/app/alembic/versions/c43313eca57d_add_document_tables.py:
```diff
@@ -0,0 +1,41 @@
+        sa.Column(
+            "object_store_url", sqlmodel.sql.sqltypes.AutoString(), nullable=False
+        ),
+        sa.Column("created_at", sa.DateTime(), nullable=False),
```
This is not appropriate to fix in this PR

jerome-white commented on backend/app/crud/document.py:
```diff
@@ -3,54 +3,52 @@
-        statement = (
-            select(Document)
-            .where(and_(
+        statement = select(Document).where(
+            and_(
                 Document.owner_id == self.owner_id,
                 Document.id == doc_id,
-            ))
+            )
         )
 
         return self.session.exec(statement).one()
```
This is not appropriate to fix in this PR

jerome-white committed:
Pre-commit formatting

jerome-white committed:
Give AWS region a valid default when testing

jerome-white committed:
Add a default region to the environment

jerome-white commented on backend/app/models/user.py:
```diff
@@ -49,8 +49,12 @@ class User(UserBase, table=True):
     items: list["Item"] = Relationship(back_populates="owner", cascade_delete=True)
-    documents: list["Document"] = Relationship(back_populates="owner", cascade_delete=True)
-    projects: list["ProjectUser"] = Relationship(back_populates="user", cascade_delete=True)
+    documents: list["Document"] = Relationship(
```
Addressable in a future PR

AkhileshNegi commented on .env.example:
```diff
@@ -61,7 +61,7 @@ DOCKER_IMAGE_FRONTEND=frontend
 AWS_ACCESS_KEY_ID=
 AWS_SECRET_ACCESS_KEY=
-AWS_DEFAULT_REGION=
+AWS_DEFAULT_REGION=ap-south-1
```
do we need this?

jerome-white replied:
Yes, there needs to be a default value that is a valid AWS region. See [Discord](https://discord.com/channels/1014768296257654865/1358755777116110946/1358759480992731408)

avirajsingh7 commented on backend/app/api/routes/documents.py:
```diff
@@ -1,44 +1,53 @@
+    skip: int = 0,
+    limit: int = 100,
 ):
     crud = DocumentCrud(session, current_user.id)
```
@jerome-white Add check for skip>0 and limit< some value (100 may be)
```
    skip: int = Query(0, ge=0),
    limit: int = Query(100, ge=1, le=100),
```

jerome-white replied:
Done: 5242ac9

jerome-white committed:
Fill now returns a list

jerome-white committed:
Test skips larger than rows

jerome-white committed:
Test skip and limit together

jerome-white committed:
Do not assume document fetch ordering

jerome-white committed:
Combination of skip and limit perform as expected

jerome-white committed:
Add parameter value checks

https://github.com/ProjectTech4DevAI/ai-platform/pull/124#discussion_r2031525677
